### PR TITLE
replay fix proposition

### PR DESCRIPTION
**Pull request by AlbinSou**
Linked to issue #1274. I think this should  be enough for the fix, however I would prefer to wait for confirmation that this does not break anything. As far as I understood, loader_data is a dictionnary containing data for the potential multiple tasks present in the current experience. So, taking the maximum length over just this dictionnary should ensure we see all the data from current task, and loop on the other task dataloader if they are of unequal size. However, we don't want to take the maximum lenght also from replay loader since it can lead to the bug described in the above issue.

I also removed the part where the length of the memory buffers were added to the loaders used for length estimation.

### CODE REVIEW

**Comment by HamedHemati** [comment]
> Linked to issue #1274. I think this should be enough for the fix, however I would prefer to wait for confirmation that this does not break anything. As far as I understood, loader_data is a dictionnary containing data for the potential multiple tasks present in the current experience. So, taking the maximum length over just this dictionnary should ensure we see all the data from current task, and loop on the other task dataloader if they are of unequal size. However, we don't want to take the maximum lenght also from replay loader since it can lead to the bug described in the above issue.
> 
> I also removed the part where the length of the memory buffers were added to the loaders used for length estimation.

Thanks @AlbinSou for the PR. I also tested similar changes in a local branch, and it worked fine. Just to be on the safe side, did you test `examples/replay.py` with very small and very large buffer sizes?

**Comment by AlbinSou** [comment]
For now I tried the online replay case. I will try some things as you suggested @HamedHemati with big buffer sizes and non-online setting. @AntonioCarta in the continual learning baselines, I think the only one using ReplayDataLoader is Cope, which automatically cuts training after a set number of iterations. But you can see that with the old replay plugin the number of iteration given by the logger (before getting cut) gets bigger and bigger. With the fix it is not the case anymore.

**Comment by AlbinSou** [comment]
For instance, when I try the old version with Split CIFAR100, 20 splits,10000 memory, bs 10, this is the number of iteration per experience that I get with the old plugin:

1st exp: 237 (task set size: 2375, memory size: 0)
2nd exp: 238 (task set size: 2375, memory size: 2375) -> so far no problem, just one more iteration
3rd exp: 475 (task set size: 2375, memory size: 4750)
4rth exp: 713 ...
5th exp: 950
6th exp: 1000

[...]

This keeps going on until memory is filled fully and then sticks to 1000 iteration (full pass on the full memory)

With the fix, I get 237 iterations in the first experience, 238 in the following ones. Maybe there is still a problem in the sense that this does not drop the last batch of current data. I traced it back to the ReplayPlugin in itself which is not passing the drop_last argument to the loader, I will change that too.

**Comment by AlbinSou** [comment]
> Thanks @AlbinSou, the change looks ok. Can you also update the docstring to describe the new behavior? [This](https://github.com/ContinualAI/avalanche/pull/1287/files#diff-791d75e9921e2a6f9380555947aafadca9a575c54bbdda8cea3102aac93153e6L337) in particular seems outdated now.
> 
> We should definitely test this thoroughly, maybe you can try to run some of the continual-learning-baselines scripts that use the ReplayDataLoader?

About the doc string, I looked at it but I don't really know what is to change. I guess I should test the oversample_small_tasks and task_balanced_loader options which only occurs when task labels are provided. But the rest of the docstring is pretty clear and already explains the current (fixed behavior). Maybe I should precise that the loader will be of the same length than the one from current task to make it clear. Also the distributed sampling option is not explained, but I honestly don't know what it is used for and I don't know how to test it either.

**Comment by AntonioCarta** [comment]
> Maybe I should precise that the loader will be of the same length than the one from current task to make it clear

Yes, you are right. Just mention that the loader uses the experience data to set the number of mini-batches.

Also, can you add a test to check that the number of iterations is correct? To avoid breaking it in the future.

**Comment by AlbinSou** [comment]
I added a test and it seems like it was well needed, because one of them fails. When I check the batch size, it should be always equal to 2 times the current task batch size (if I set batch_size=current_task_bs and batch_size_mem=current_task_bs). However, it's not the case in some cases. I will investigate that but if you have an idea in the meantime it would be welcomed. I suspect the looping onto memory loader might not be effective if done at all, since this bug happens only for 3 tests, where the memory is smaller than the current task set.

**Comment by AntonioCarta** [comment]
Maybe it's happening only for the last minibatch if you don't have a dataset divisible by `batch_size`?

**Comment by AlbinSou** [comment]
> Maybe it's happening only for the last minibatch if you don't have a dataset divisible by `batch_size`?

No because I am using drop_last=True. And also everytime the batch is not a random size, it's either batch_size + batch_size_mem or just batch_size

**Comment by AlbinSou** [comment]
Ok I fixed this it's just the normal behavior when the option oversample_small_task is not used, but it is used by default inside of the ReplayPlugin. Activating it makes the test pass.

**Comment by AntonioCarta** [comment]
Great! Can I merge this or are you planning any other changes?

**Comment by AlbinSou** [comment]
> Great! Can I merge this or are you planning any other changes?

You can merge, no other changes planned for now
